### PR TITLE
Add school info to pd enrollment and use DSCO take 2

### PR DESCRIPTION
**Pull request by drizco**
This pr adds the new `SchoolDataInputs` component to the workshop enrollment form. Since the school info component uses the design system form components, the entire form needed to be updated

INTRO WORKSHOP BEFORE:
localhost-studio code org_3000_pd_workshops_1_enroll (7)

INTRO WORKSHOP AFTER:

localhost-studio code org_3000_pd_workshops_1_enroll

DEEP DIVE WORKSHOP BEFORE:
localhost-studio code org_3000_pd_workshops_1_enroll (2)

DEEP DIVE WORKSHOP AFTER:
localhost-studio code org_3000_pd_workshops_1_enroll (1)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/lib/school_info_deduplicator.rb [hunk 1]
```diff
@@ -28,26 +28,12 @@ def get_duplicate_school_info(school_info_attr)
   def process_school_info_attributes(school_info_attr)
     attr = school_info_attr.symbolize_keys
 
-    # Names of state and zip fields change between form fields and SchoolInfo class
```
Comment (bethanyaconnor) [inline dashboard/lib/school_info_deduplicator.rb]: Why can this stuff be removed now?

File: dashboard/app/controllers/api/v1/user_school_infos_controller.rb [hunk 2]
```diff
@@ -32,8 +32,6 @@ def update
         else
           # VALIDATION_COMPLETE is passed when the school_id does not exist to check
           # form for completeness; specifically, school name is required.
-          # If school_id does not exist, ncesSchoolId is set to -1 when the checkbox
-          # for school not found is clicked.
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/api/v1/user_school_infos_controller.rb:36]: Is this comment no longer true?

File: apps/src/code-studio/pd/form_components/QuestionsTable.jsx [hunk 3]
```diff
@@ -97,7 +100,7 @@ export default class QuestionsTable extends React.Component {
     };
 
     return (
-      <FormGroup>
+      <FormGroup bsClass={this.props.className}>
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/form_components/QuestionsTable.jsx:103]: for my own learning, what is the prop `bsClass`?

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx [hunk 4]
```diff
@@ -1,174 +1,193 @@
 /*
  * Form to create a workshop enrollment
  */
-import React, {useState} from 'react';
-import {FormGroup, ControlLabel, HelpBlock, Alert} from 'react-bootstrap'; // eslint-disable-line no-restricted-imports
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:5]: love getting rid of a disabled linter directive 🤩

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx [hunk 5]
```diff
@@ -1,174 +1,193 @@
...
+  DISTRICT,
+  EXPLAIN,
+  GRADES_TEACHING,
+  INTRO,
+  NOT_TEACHING,
+  OTHER,
+  SUBMISSION_STATUSES,
+} from './constants';
 
-const CSF = 'CS Fundamentals';
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:24]: It looks like all these constants were moved to `constants.ts`?

File: apps/src/schoolInfo/utils/buildSchoolData.ts [hunk 6]
```diff
@@ -38,25 +34,17 @@ export function buildSchoolData({
     schoolId === NonSchoolOptions.NO_SCHOOL_SETTING
   ) {
     return {
-      user: {
```
Comment (bethanyaconnor) [inline apps/src/schoolInfo/utils/buildSchoolData.ts:41]: Can you explain why this change is necessary? I have a guess but I'm not sure!

File: apps/src/code-studio/pd/form_components/QuestionsTable.jsx (hunk 3 above)
Comment (drizco) [reply apps/src/code-studio/pd/form_components/QuestionsTable.jsx:103]: bootstrap class! I needed a way to target just the inputs in this QuestionsTable component that was built with bootstrap components

File: apps/src/schoolInfo/utils/buildSchoolData.ts (hunk 6 above)
Comment (drizco) [reply apps/src/schoolInfo/utils/buildSchoolData.ts:41]: when I first created the `buildSchoolData` helper function, I thought it was only going to be used with `PATCH /api/v1/user_school_infos` where the params are extracted in the `user_school_infos_controller`. un-nesting the attributes made it reusable in the `POST /api/v1/pd/workshop/:workshopId/enrollments` route

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx (hunk 5 above)
Comment (drizco) [reply apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:24]: yep! it was getting pretty cluttered

File: dashboard/app/controllers/api/v1/user_school_infos_controller.rb (hunk 2 above)
Comment (drizco) [reply dashboard/app/controllers/api/v1/user_school_infos_controller.rb:36]: it's no longer true. we use the absence of school_id to know that it's not an nces school. we're not storing "-1" in the db either:
Screenshot 2024-12-05 at 10 51 48 AM

File: dashboard/lib/school_info_deduplicator.rb (hunk 1 above)
Comment (drizco) [reply dashboard/lib/school_info_deduplicator.rb]: it started with removing line 33, which is just translating the shape of the request. since `buildSchoolData` sends `zip` and not `school_zip` this translation was no longer necessary. then since we don't care about district, address, etc for non nces schools, I decided to clean this up as well. of course that meant more tests to update... I think it's still a good change, but I could wait on it and do it in the school_info cleanup ticket if that's preferred

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx (hunk 5 above)
Comment (bethanyaconnor) [reply apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:24]: Yeah, I like the new constants file! It's a nice balance of decluttering these files without cluttering the shared constants (but also allows constants to be moved to the shared constants if that makes sense!)

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx [hunk 7]
```diff
@@ -1,174 +1,193 @@
...
-  workshop_enrollment_status: string;
   account_exists: boolean;
-  sign_up_url: string;
   cancel_url: string;
+  sign_up_url: string;
+  workshop_enrollment_status: string;
 };
 
-type EnrollFormProps = EnrollFormState & {
+type EnrollFormProps = {
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:96]: Why did so many things need to be added? Is this cleanup or a new need? (If it's cleanup, I think it makes sense in this PR, just want to understand!)

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx [hunk 8]
```diff
@@ -352,102 +362,129 @@ export default function EnrollForm(props: EnrollFormProps) {
         }
       );
       setIsSubmitting(false);
-      if (response.ok) {
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:355]: Why remove this check??

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx [hunk 9]
```diff
@@ -352,102 +362,129 @@ export default function EnrollForm(props: EnrollFormProps) {
...
     ];
 
     if (!props.email) {
-      requiredFields.push('confirm_email');
+      fields.push('confirm_email');
     }
 
     if (props.workshop_course === CSF) {
-      requiredFields.push('role', 'grades_teaching');
+      fields.push('role', 'grades_teaching');
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:442]: Can you explain what this change is doing?

File: apps/src/schoolInfo/utils/buildSchoolData.ts (hunk 6 above)
Comment (bethanyaconnor) [reply apps/src/schoolInfo/utils/buildSchoolData.ts:41]: Is this function not used in sign up?

File: apps/style/code-studio/pd.scss [hunk 10]
```diff
@@ -29,8 +29,7 @@
 // TODO: Remove this once Dashboard is upgraded to Bootstrap 3
```
Comment (bethanyaconnor) [inline apps/style/code-studio/pd.scss:29]: 😭

File: dashboard/lib/school_info_deduplicator.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/lib/school_info_deduplicator.rb]: Given some things are still using the old school info inputs, do they not still need this code path?

**Review by bethanyaconnor — COMMENTED** [review]
If we get a chance, I'd be interested in a live voice-over of the changes to `enroll_form.tsx` (and, to a lesser degree, the accompanying test). Enough things are being moved/refactored that I'm struggling to understand what's changed versus moved.

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx (hunk 7 above)
Comment (drizco) [reply apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:96]: I needed a way to render the component with some initial state in order to avoid an even more extensive refactor of the test file. the added optional props are not actually passed to the component where it's implemented in `workshop_enroll.jsx`, only in the tests

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx (hunk 8 above)
Comment (drizco) [reply apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:355]: ah, this was tricky! any validation errors are returned in the response json, so it needs to be parsed regardless of the response status and handled in `onSubmissionComplete`. this was a bug I accidentally introduced when I first refactored the component

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx (hunk 9 above)
Comment (drizco) [reply apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:442]: I've changed a function for getting required fields to a memoized property derived from state. then on lines 432 to 451, I'm making the "explain" textareas required if their corresponding options are checked. this was missing from the original implementation

File: apps/src/schoolInfo/utils/buildSchoolData.ts (hunk 6 above)
Comment (drizco) [reply apps/src/schoolInfo/utils/buildSchoolData.ts:41]: it's not, only in the `updateSchoolInfo` helper method that calls the api

File: dashboard/lib/school_info_deduplicator.rb (hunk 1 above)
Comment (drizco) [reply dashboard/lib/school_info_deduplicator.rb]: yeah, that's a fair point! I'll hold off on this change and note it for the cleanup pr

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx (hunk 7 above)
Comment (drizco) [reply apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:96]: I will add a comment explaining this

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx [hunk 11]
```diff
@@ -505,209 +534,262 @@ export default function EnrollForm(props: EnrollFormProps) {
...
-              }
-              validationState={
-                Object.prototype.hasOwnProperty.call(formErrors, 'email')
-                  ? VALIDATION_STATE_ERROR
-                  : null
-              }
-              errorMessage={formErrors.email}
+            <Label
+              errorMessage={formErrors.school_info}
+              htmlFor="school_info"
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx]: I _think_ this needs to be the ID of the input -- can you export the ID from the component file?

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx (hunk 7 above)
Comment (bethanyaconnor) [reply apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:96]: Did you make a ticket to move the test to RTL?

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx [hunk 12]
```diff
@@ -1,174 +1,199 @@
...
+  grades_teaching?: string[];
+  csf_intro_intent?: string;
+  attended_csf_intro_workshop?: string;
+  years_teaching?: string;
+  years_teaching_cs?: string;
+  taught_ap_before?: string;
+  planning_to_teach_ap?: string;
 };
 
+const Label = ({
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:124]: Sorry, should've asked for this this morning, but can you add a comment detailing why this is needed?

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx [hunk 13]
```diff
@@ -505,209 +540,262 @@ export default function EnrollForm(props: EnrollFormProps) {
...
-      default:
-        return [];
-    }
-  };
-
-  const roles = getRoles(props.workshop_course);
-
   return (
-    <form id="enroll-form">
+    <form id={styles.enroll_form}>
```
Comment (bethanyaconnor) [inline apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:556]: TIL that you can use this syntax for IDs as well!

**Review by bethanyaconnor — APPROVED** [review]
Had another comment request but LGTM. Thanks for walking me through this this morning!

File: apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx (hunk 7 above)
Comment (drizco) [reply apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx:96]: just did! https://codedotorg.atlassian.net/browse/ACQ-2917
